File: python/mhc/metrics.py
"""
Stability metrics for analyzing residual mixing matrices.

These metrics quantify how a matrix amplifies signals during forward/backward
propagation through a neural network layer.

Key insight from the mHC paper:
- Unconstrained matrices (HC) can have unbounded gains, causing signal explosion
- Doubly stochastic matrices (mHC) have all gains bounded by 1, ensuring stability

Metrics:
- forward_gain: Worst-case signal amplification in forward pass (max row sum)
- backward_gain: Worst-case gradient amplification in backward pass (max column sum)
- spectral_norm: Largest singular value (general operator norm)

For doubly stochastic matrices, all three equal exactly 1.

Author: Subhadip Mitra <contact@subhadipmitra.com>
Based on DeepSeek's mHC paper: https://arxiv.org/abs/2512.24880
"""

import numpy as np


def eigenvalues_sorted(matrix: np.ndarray) -> np.ndarray:
    """
    Compute eigenvalues sorted by magnitude (descending).

    For doubly stochastic matrices:
    - The largest eigenvalue is always 1 (eigenvector = all-ones)
    - All other eigenvalues have |λ| <= 1
    - Products of DS matrices push subdominant eigenvalues toward 0

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        Array of eigenvalues sorted by |λ| descending

    Example:
        >>> eigs = eigenvalues_sorted(np.eye(4))
        >>> np.allclose(eigs, [1, 1, 1, 1])
        True
    """
    eigenvalues = np.linalg.eigvals(matrix)
    # Sort by magnitude descending
    idx = np.argsort(np.abs(eigenvalues))[::-1]
    return eigenvalues[idx]


def second_largest_eigenvalue_magnitude(matrix: np.ndarray) -> float:
    """
    Compute the magnitude of the second-largest eigenvalue.

    This is the key quantity for understanding convergence in products of
    doubly stochastic matrices. For DS matrices:
    - λ_1 = 1 always (the Perron-Frobenius eigenvalue)
    - |λ_2| < 1 for generic DS matrices (spectral gap)
    - After L multiplications, |λ_2|^L → 0, causing convergence to uniform matrix

    The rate of convergence to the uniform matrix 1/n is governed by |λ_2|.
    Larger |λ_2| = slower convergence = more information preserved.

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        |λ_2|, the second-largest eigenvalue magnitude

    Example:
        >>> second_largest_eigenvalue_magnitude(np.eye(4))
        1.0
        >>> # For a doubly stochastic matrix that's not a permutation:
        >>> P = np.array([[0.5, 0.5], [0.5, 0.5]])  # Uniform matrix
        >>> second_largest_eigenvalue_magnitude(P) < 0.01
        True
    """
    eigs = eigenvalues_sorted(matrix)
    if len(eigs) < 2:
        return 0.0
    return float(np.abs(eigs[1]))


def distance_from_uniform(matrix: np.ndarray) -> float:
    """
    Compute Frobenius norm distance from the uniform averaging matrix.

    The uniform matrix U has all entries equal to 1/n, where n is the matrix size.
    For doubly stochastic matrices, repeated multiplication converges to U.

    This metric directly shows how much "information mixing" has occurred:
    - Distance = 0 means complete convergence to uniform (all information averaged)
    - Distance = sqrt(n-1) for identity matrix (no mixing)
    - Distance decreases with each DS matrix multiplication

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        ||M - U||_F, the Frobenius norm distance from uniform matrix

    Example:
        >>> distance_from_uniform(np.eye(4))  # Identity is far from uniform
        1.7320508075688772
        >>> distance_from_uniform(np.ones((4, 4)) / 4)  # Uniform matrix
        0.0
    """
    n = matrix.shape[0]
    uniform = np.ones((n, n)) / n
    return float(np.linalg.norm(matrix - uniform, 'fro'))


def forward_gain(matrix: np.ndarray) -> float:
    """
    Compute maximum absolute row sum (worst-case signal amplification).

    This is the infinity norm ||M||_∞, which equals the maximum amplification
    a unit input vector can experience in the forward pass:
        ||Mx||_∞ <= ||M||_∞ * ||x||_∞

    For a doubly stochastic matrix, all row sums equal 1, so forward_gain = 1.
    For unconstrained matrices, can be arbitrarily large.

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        Maximum absolute row sum: max_i |sum_j M[i,j]|

    Example:
        >>> forward_gain(np.eye(4))
        1.0
        >>> forward_gain(np.ones((4, 4)))
        4.0
    """
    return float(np.abs(matrix.sum(axis=1)).max())


def backward_gain(matrix: np.ndarray) -> float:
    """
    Compute maximum absolute column sum (worst-case gradient amplification).

    This is the one norm ||M||_1, which equals the maximum amplification
    a gradient vector can experience in the backward pass:
        ||M^T g||_1 <= ||M||_1 * ||g||_1

    For a doubly stochastic matrix, all column sums equal 1, so backward_gain = 1.
    For unconstrained matrices, can be arbitrarily large.

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        Maximum absolute column sum: max_j |sum_i M[i,j]|

    Example:
        >>> backward_gain(np.eye(4))
        1.0
        >>> backward_gain(np.ones((4, 4)))
        4.0
    """
    return float(np.abs(matrix.sum(axis=0)).max())


def spectral_norm(matrix: np.ndarray, iterations: int = 20) -> float:
    """
    Estimate spectral norm (largest singular value) via power iteration.

    The spectral norm ||M||_2 is the maximum amplification of a unit vector
    in the L2 sense. For any input x with ||x||_2 = 1:
        ||Mx||_2 <= ||M||_2

    For doubly stochastic matrices, spectral_norm <= 1.

    Algorithm (power iteration):
    1. Start with random unit vector v
    2. Iterate: v = M @ v / ||M @ v||
    3. Estimate: ||M @ v|| converges to largest singular value

    Args:
        matrix: Input matrix of shape (n, n)
        iterations: Number of power iterations (20 is usually sufficient)

    Returns:
        Estimated spectral norm (largest singular value)

    Example:
        >>> spectral_norm(np.eye(4))  # doctest: +ELLIPSIS
        1.0...
        >>> spectral_norm(2 * np.eye(4))  # doctest: +ELLIPSIS
        2.0...
    """
    n = matrix.shape[0]

    # Initialize with random unit vector
    rng = np.random.default_rng(42)  # Fixed seed for reproducibility
    v = rng.standard_normal(n)
    v = v / np.linalg.norm(v)

    for _ in range(iterations):
        # Power iteration: v = M @ v, then normalize
        w = matrix @ v
        norm = np.linalg.norm(w)
        if norm < 1e-10:
            return 0.0
        v = w / norm

    # Final estimate: ||M @ v||
    return float(np.linalg.norm(matrix @ v))


def compute_all_metrics(matrix: np.ndarray) -> dict:
    """
    Compute all stability metrics for a matrix.

    This is the main function for analyzing residual mixing matrices.
    It returns all metrics needed to assess training stability.

    Args:
        matrix: Input matrix of shape (n, n)

    Returns:
        Dict containing:
        - spectral_norm: Largest singular value
        - forward_gain: Max absolute row sum
        - backward_gain: Max absolute column sum
        - row_sum_max_dev: Max deviation of row sums from 1
        - col_sum_max_dev: Max deviation of column sums from 1
        - min_entry: Minimum matrix entry

    Example:
        >>> metrics = compute_all_metrics(np.eye(4))
        >>> metrics['forward_gain']
        1.0
        >>> metrics['backward_gain']
        1.0
    """
    row_sums = matrix.sum(axis=1)
    col_sums = matrix.sum(axis=0)

    # Compute eigenvalues for convergence analysis
    eigs = eigenvalues_sorted(matrix)
    largest_eig_mag = float(np.abs(eigs[0])) if len(eigs) > 0 else 0.0
    second_eig_mag = float(np.abs(eigs[1])) if len(eigs) > 1 else 0.0

    return {
        'spectral_norm': spectral_norm(matrix),
        'forward_gain': float(np.abs(row_sums).max()),
        'backward_gain': float(np.abs(col_sums).max()),
        'row_sum_max_dev': float(np.abs(row_sums - 1).max()),
        'col_sum_max_dev': float(np.abs(col_sums - 1).max()),
        'min_entry': float(matrix.min()),
        'largest_eigenvalue_mag': largest_eig_mag,
        'second_eigenvalue_mag': second_eig_mag,
        'distance_from_uniform': distance_from_uniform(matrix),
    }

File: react-demo/src/lib/metrics.ts
/**
 * Stability metrics for analyzing residual mixing matrices.
 * TypeScript port of the Python implementation.
 */

import type { Metrics } from './types';

/**
 * Compute eigenvalues using QR iteration (for small matrices).
 * Returns eigenvalues sorted by magnitude (descending).
 */
export function eigenvaluesSorted(matrix: number[][]): { real: number; imag: number }[] {
  const n = matrix.length;

  // Make a copy to work with
  let A = matrix.map(row => [...row]);

  // QR iteration - converges to Schur form
  const maxIters = 100;
  for (let iter = 0; iter < maxIters; iter++) {
    // QR decomposition using Gram-Schmidt
    const { Q, R } = qrDecomposition(A);
    // A = R * Q (reverse multiply)
    A = matrixMultiply(R, Q);
  }

  // Extract eigenvalues from diagonal (for real eigenvalues)
  // For 2x2 blocks on diagonal, compute complex eigenvalues
  const eigenvalues: { real: number; imag: number }[] = [];

  let i = 0;
  while (i < n) {
    if (i === n - 1 || Math.abs(A[i + 1][i]) < 1e-10) {
      // Real eigenvalue on diagonal
      eigenvalues.push({ real: A[i][i], imag: 0 });
      i++;
    } else {
      // 2x2 block - compute complex eigenvalue pair
      const a = A[i][i];
      const b = A[i][i + 1];
      const c = A[i + 1][i];
      const d = A[i + 1][i + 1];

      const trace = a + d;
      const det = a * d - b * c;
      const discriminant = trace * trace - 4 * det;

      if (discriminant >= 0) {
        const sqrtDisc = Math.sqrt(discriminant);
        eigenvalues.push({ real: (trace + sqrtDisc) / 2, imag: 0 });
        eigenvalues.push({ real: (trace - sqrtDisc) / 2, imag: 0 });
      } else {
        const realPart = trace / 2;
        const imagPart = Math.sqrt(-discriminant) / 2;
        eigenvalues.push({ real: realPart, imag: imagPart });
        eigenvalues.push({ real: realPart, imag: -imagPart });
      }
      i += 2;
    }
  }

  // Sort by magnitude descending
  eigenvalues.sort((a, b) => {
    const magA = Math.sqrt(a.real * a.real + a.imag * a.imag);
    const magB = Math.sqrt(b.real * b.real + b.imag * b.imag);
    return magB - magA;
  });

  return eigenvalues;
}

/**
 * QR decomposition using modified Gram-Schmidt.
 */
function qrDecomposition(A: number[][]): { Q: number[][]; R: number[][] } {
  const n = A.length;
  const Q: number[][] = Array(n).fill(null).map(() => Array(n).fill(0));
  const R: number[][] = Array(n).fill(null).map(() => Array(n).fill(0));

  // Copy columns of A
  const columns: number[][] = [];
  for (let j = 0; j < n; j++) {
    columns.push(A.map(row => row[j]));
  }

  for (let j = 0; j < n; j++) {
    let v = [...columns[j]];

    // Subtract projections onto previous q vectors
    for (let i = 0; i < j; i++) {
      const qi = Q.map(row => row[i]);
      R[i][j] = dotProduct(qi, columns[j]);
      v = v.map((val, idx) => val - R[i][j] * qi[idx]);
    }

    R[j][j] = Math.sqrt(dotProduct(v, v));

    if (R[j][j] > 1e-10) {
      for (let i = 0; i < n; i++) {
        Q[i][j] = v[i] / R[j][j];
      }
    } else {
      // Handle zero column
      for (let i = 0; i < n; i++) {
        Q[i][j] = i === j ? 1 : 0;
      }
    }
  }

  return { Q, R };
}

/**
 * Matrix multiplication helper.
 */
function matrixMultiply(A: number[][], B: number[][]): number[][] {
  const n = A.length;
  const result: number[][] = Array(n).fill(null).map(() => Array(n).fill(0));

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      for (let k = 0; k < n; k++) {
        result[i][j] += A[i][k] * B[k][j];
      }
    }
  }

  return result;
}

/**
 * Dot product of two vectors.
 */
function dotProduct(a: number[], b: number[]): number {
  return a.reduce((sum, val, i) => sum + val * b[i], 0);
}

/**
 * Get the magnitude of the second-largest eigenvalue.
 * This determines how fast the matrix product converges to uniform.
 */
export function secondLargestEigenvalueMagnitude(matrix: number[][]): number {
  const eigs = eigenvaluesSorted(matrix);
  if (eigs.length < 2) return 0;
  const e = eigs[1];
  return Math.sqrt(e.real * e.real + e.imag * e.imag);
}

/**
 * Get the magnitude of the largest eigenvalue.
 */
export function largestEigenvalueMagnitude(matrix: number[][]): number {
  const eigs = eigenvaluesSorted(matrix);
  if (eigs.length < 1) return 0;
  const e = eigs[0];
  return Math.sqrt(e.real * e.real + e.imag * e.imag);
}

/**
 * Compute Frobenius norm distance from the uniform averaging matrix.
 * Shows how much "information mixing" has occurred.
 */
export function distanceFromUniform(matrix: number[][]): number {
  const n = matrix.length;
  const uniformVal = 1 / n;
  let sumSq = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const diff = matrix[i][j] - uniformVal;
      sumSq += diff * diff;
    }
  }

  return Math.sqrt(sumSq);
}

/**
 * Compute forward gain (max absolute row sum).
 */
export function forwardGain(matrix: number[][]): number {
  const rowSums = matrix.map((row) => row.reduce((a, b) => a + b, 0));
  return Math.max(...rowSums.map(Math.abs));
}

/**
 * Compute backward gain (max absolute column sum).
 */
export function backwardGain(matrix: number[][]): number {
  const n = matrix.length;
  const colSums: number[] = Array(n).fill(0);

  for (let j = 0; j < n; j++) {
    for (let i = 0; i < n; i++) {
      colSums[j] += matrix[i][j];
    }
  }

  return Math.max(...colSums.map(Math.abs));
}

/**
 * Estimate spectral norm via power iteration.
 */
export function spectralNorm(
  matrix: number[][],
  iterations: number = 20
): number {
  const n = matrix.length;

  // Random unit vector (using fixed seed for reproducibility)
  let v = Array(n)
    .fill(null)
    .map((_, i) => Math.sin(i + 1));
  let norm = Math.sqrt(v.reduce((a, b) => a + b * b, 0));
  v = v.map((x) => x / norm);

  for (let iter = 0; iter < iterations; iter++) {
    // w = M @ v
    const w = Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        w[i] += matrix[i][j] * v[j];
      }
    }

    norm = Math.sqrt(w.reduce((a, b) => a + b * b, 0));
    if (norm < 1e-10) return 0;
    v = w.map((x) => x / norm);
  }

  // Final norm estimate
  const Mv = Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      Mv[i] += matrix[i][j] * v[j];
    }
  }
  return Math.sqrt(Mv.reduce((a, b) => a + b * b, 0));
}

/**
 * Compute row sum max deviation from 1.
 */
export function rowSumMaxDev(matrix: number[][]): number {
  const rowSums = matrix.map((row) => row.reduce((a, b) => a + b, 0));
  return Math.max(...rowSums.map((s) => Math.abs(s - 1)));
}

/**
 * Compute column sum max deviation from 1.
 */
export function colSumMaxDev(matrix: number[][]): number {
  const n = matrix.length;
  const colSums: number[] = Array(n).fill(0);

  for (let j = 0; j < n; j++) {
    for (let i = 0; i < n; i++) {
      colSums[j] += matrix[i][j];
    }
  }

  return Math.max(...colSums.map((s) => Math.abs(s - 1)));
}

/**
 * Find minimum entry in matrix.
 */
export function minEntry(matrix: number[][]): number {
  let min = Infinity;
  for (const row of matrix) {
    for (const val of row) {
      if (val < min) min = val;
    }
  }
  return min;
}

/**
 * Compute all metrics for a matrix.
 */
export function computeAllMetrics(matrix: number[][]): Metrics {
  return {
    spectralNorm: spectralNorm(matrix),
    forwardGain: forwardGain(matrix),
    backwardGain: backwardGain(matrix),
    rowSumMaxDev: rowSumMaxDev(matrix),
    colSumMaxDev: colSumMaxDev(matrix),
    minEntry: minEntry(matrix),
    largestEigenvalueMag: largestEigenvalueMagnitude(matrix),
    secondEigenvalueMag: secondLargestEigenvalueMagnitude(matrix),
    distanceFromUniform: distanceFromUniform(matrix),
  };
}
